rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,168 @@
+    metrics are transformed before being sent, ultimately metrics accounts for two differents kind of payload:
+- The count, gauge and rate series kind of payload, sent to `/api/v1/series` using the [JSON schema officially
+  documented](https://docs.datadoghq.com/api/latest/metrics) with few undocumented [additional
+  fields](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/series.go#L45-L57), but this align very well
+  with the existing `datadog_metrics` sinks.
+- The sketches kind of payload, sent to `/api/beta/sketches` and serialized as protobuf as shown
+  [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/serializer/serializer.go#L315-L338) (it ultimately lands
+  [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/sketch_series.go#L103-L269)). Public `.proto`
+  definition can be found
+  [here](https://github.com/DataDog/agent-payload/blob/master/proto/metrics/agent_payload.proto#L47-L81).
+
+Vector has a nice description of its [metrics data
+model](https://vector.dev/docs/about/under-the-hood/architecture/data-model/metric/) and a [concise enum for
+representing it](https://github.com/timberio/vector/blob/master/lib/vector-core/src/event/metric.rs#L135-L169).
+
+
+The implementation would the consist in:
+
+- Implement in the `datadog_agent` source the ability to proxy all known route (see above) as-is directly to Datadog
+  (this will require an new settings in the Vector source, something like `unused_payload_site: us3.datadoghq.com`). At
+  the beginning it will also proxify `/api/v1/series` & `/api/beta/sketches`.
+- Handle the `/api/v1/series` route (based on both the offical [API](https://docs.datadoghq.com/api/latest/metrics/) and
+  the [Datadog Agent itself](https://github.com/DataDog/datadog-agent/blob/main/pkg/forwarder/telemetry.go#L20-L31)) to
+  cover every metric type handled by this endpoint (count, gauge and rate) and:
+    - Add support for missing fields in the `datadog_metrics` sinks
+    - Align tag manipulation (Datadog allows `key:foo` & `key:bar`, and the tags are encoded as a list whereas Vector
+      encode those as a map).
+    - Overall this is fairly straighforward except for the tagging issue
+- Handle the `/api/beta/sketches` route in the `datadog_agent` source to support sketches/distribution encoded using
+  protobuf
+   - Revamp the distribution metrics in the `datadog_metrics` sink to use sketches and the associated endpoint will
+     probably be a good idea.
+   - The sketches the agent ships is based on this [paper](http://www.vldb.org/pvldb/vol12/p2195-masson.pdf) whereas
+     Vector uses what's called a summary inside the Agent, implementing the complete DDSketch support in Vector is
+     probably a good idea as sketches have convenient properties for wide consistent aggregation and limited error. To
+     support smooth migration, full DDsktech support is mandatory, as customers that emit distribution metric from
```

Comment:
Found this Rust port of DDSketch - https://github.com/mheffner/rust-sketches-ddsketch
@tobz  is one of the contributors actually :)

Reply:
Nice :) I will check that out !

Reply:
And also @tobz - forked, https://github.com/metrics-rs/rust-sketches-ddsketch

Reply:
Just deleted that fork, since I only used it for a PR to the actual crate. :)

Reply:
Wasn't sure if it was "adopted" or getting adopted by metrics-rs 👍  @tobz